x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenizer.java

```diff
@@ -93,33 +93,46 @@ public BasicTokenizer(boolean isLowerCase, boolean isTokenizeCjkChars) {
                 continue;
             }
 
-            if (neverSplit.contains(token)) {
-                processedTokens.add(token);
-                continue;
+            if (isLowerCase) {
+                token = token.toLowerCase(Locale.ROOT);
             }
-
-            // At this point text has been tokenized by whitespace
-            // but one of the special never split tokens could be adjacent
-            // to one or more punctuation characters.
-            List<String> splitOnCommonTokens = splitOnPredicate(token, BasicTokenizer::isCommonPunctuation);
-            for (String splitOnCommon : splitOnCommonTokens) {
-                if (neverSplit.contains(splitOnCommon)) {
-                    processedTokens.add(splitOnCommon);
-                } else {
-                    if (isLowerCase) {
-                        splitOnCommon = splitOnCommon.toLowerCase(Locale.ROOT);
-                    }
-                    if (isStripAccents) {
-                        splitOnCommon = stripAccents(splitOnCommon);
-                    }
-                    processedTokens.addAll(splitOnPunctuation(splitOnCommon));
-                }
+            if (isStripAccents) {
+                token = stripAccents(token);
             }
+            processedTokens.addAll(splitOnPunctuation(token));
         }
 
         return processedTokens;
     }
 
+    private List<String> mergeNeverSplitTokens(List<String> tokens) {
+        if (neverSplitTokenTrieRoot.isLeaf()) {
+            return tokens;
+        }
+        List<String> mergedTokens = new ArrayList<>(tokens.size());
+        List<String> matchingTokens = new ArrayList<>();
+        TokenTrieNode current = neverSplitTokenTrieRoot;
+        for (String token : tokens) {
+            TokenTrieNode childNode = current.getChild(token);
+            if (childNode == null) {
+                if (current != neverSplitTokenTrieRoot) {
+                    mergedTokens.addAll(matchingTokens);
+                    matchingTokens = new ArrayList<>();
+                    current = neverSplitTokenTrieRoot;
+                }
+                mergedTokens.add(token);
+            } else if (childNode.isLeaf()) {
+                matchingTokens.add(token);
+                mergedTokens.add(String.join("", matchingTokens));
+                current = neverSplitTokenTrieRoot;
```

Comment:
Also clear the matchingTokens here otherwise if you had 2 matches in a row `matchingTokens` would accumulate both. 

`matchingTokens = new ArrayList<>();`

Reply:
Very well spotted! I'll also add test cases that catch this.